include/wil/cppwinrt.h

```diff
@@ -403,6 +403,89 @@ namespace wil
+    // the IMyThing2 interface should be supported.
+    struct Version2 { static bool IsEnabled(); };
+
+    // Declare implementation class which conditionally supports IMyThing2.
+    struct MyThing : wil::winrt_conditionally_implements<MyThingT<MyThing>,
+                         Version2, IMyThing2>
+    {
+        // implementation goes here
+    };
+
+    ~~~~
+
+    If `Version2::IsEnabled()` returns `false`, then the `QueryInterface`
+    for `IMyThing2` will fail.
+
+    Any interface not listed as conditional is assumed to be enabled unconditionally.
+
+    You can add additional Version / Interface pairs to the template parameter list.
+    Interfaces may be conditionalized on at most one Version class. If you need a
+    complex conditional, create a new helper class.
+
+    ~~~~
+    // Helper class for testing two Versions.
+    struct Version2_or_greater {
+        static bool IsEnabled() { return Version2::IsEnabled() || Version3::IsEnabled(); }
+    };
+
+    // This implementation supports IMyThing2 if either Version2 or Version3 is enabled,
+    // and supports IMyThing3 only if Version3 is enabled.
+    struct MyThing : wil::winrt_conditionally_implements<MyThingT<MyThing>,
+    Version2_or_greater, IMyThing2, Version3, IMyThing3>
+    {
+        // implementation goes here
+    };
+    ~~~~
+    */
+    template<typename Implements, typename... Rest>
+    struct winrt_conditionally_implements : Implements
+    {
+        using Implements::Implements;
+
+        void* find_interface(winrt::guid const& iid) const noexcept override
+        {
+            static_assert(sizeof...(Rest) % 2 == 0, "Extra template parameters should come in groups of two");
+            if (is_enabled<0, std::tuple<Rest...>>(iid))
+            {
+                return Implements::find_interface(iid);
+            }
+            return nullptr;
+        }
+
+    private:
+        template<std::size_t index, typename Tuple>
+        static bool is_enabled(winrt::guid const& iid)
+        {
+            if constexpr (index >= std::tuple_size_v<Tuple>)
+            {
+                return true;
+            }
+            else
+            {
+                check_no_duplicates<1, index + 1, Tuple>();
+                return (iid == winrt::guid_of<std::tuple_element_t<index + 1, Tuple>>()) ?
+                    std::tuple_element_t<index, Tuple>::IsEnabled() :
+                    is_enabled<index + 2, Tuple>(iid);
+            }
+        }
+
+        template<std::size_t index, std::size_t upto, typename Tuple>
+        static constexpr void check_no_duplicates()
+        {
+            if constexpr (index < upto)
+            {
+                static_assert(!std::is_same_v<std::tuple_element_t<index, Tuple>, std::tuple_element_t<upto, Tuple>>,
+                    "Duplicate interfaces found in winrt_conditionally_implements");
+                check_no_duplicates<index + 2, upto, Tuple>();

```

Comment:
Feature creep - what if the "is enabled" method is constexpr? Is there a way to exclude the vtable entirely by filtering the `Rest...` somehow in that case?

Reply:
In that case, you don't need `winrt_conditionally_implements` at all.

```cpp
    struct TestClass : wil::winrt_conditionally_implements<
        winrt::implements<TestClass, std::conditional_t<also_closable, IClosable, std::true_type> >
```

If you mean that the template could autoconvert constexpr IsEnabled() into std::conditional_t, um, yeah, nice feature request.